src/dbnode/storage/index/fields_terms_iterator.go

```diff
@@ -96,6 +109,45 @@ func (fti *fieldsAndTermsIter) Reset(s segment.Segment, opts fieldsAndTermsItera
 		return err
 	}
 	fti.fieldIter = fiter
+	// If need to restrict by query, run the query on the segment first.
+	if query := opts.restrictByQuery; query != nil {
+		var (
+			success bool
+		)
+		reader, err := fti.seg.Reader()
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			if !success {
```

Comment:
I don't think `success` is set anywhere

Reply:
Good catch, ta - will update the var and set to true when we go try to close the reader.